### PR TITLE
[ISSUE#4178] When there are a large number of errors in the HTTP target, RocketMQ consumption is slow

### LINKED ISSUE #4178: [Bug] When there are a large number of errors in the HTTP target, RocketMQ consumption is slow
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/eventmesh/issues?q=is%3Aissue) and found no similar issues.

### Environment

Windows

### EventMesh version

master

### What happened

when https target errors. EM retry 3 times.  the offset will update after end retry.
But rocketmq process queue is 1000.
in this process, rocketmq consumption is slow

### How to reproduce

1、use a error https target
2、send 10000 events to em

### Debug logs

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by yanrongzhen**
Fixes #4178 

### Modifications
Provide MQ-storage-based retry strategy in http protocol.
Add retry, rocketmq-retry modules.

### CODE REVIEW

**Comment by yanrongzhen** [comment]
> > Could you explain how this PR solves the problem described in issue #4178? Because most of the code I see seems to be a unified and abstract refactoring of the code related to the grpc, http, and tcp modules.
> > 能否解释下该PR是怎么解决issue #4178所描述的问题？因为我看到的代码好像大部分是给grpc、http、tcp三个模块相关代码做统一和抽象的重构工作。
> 
> Waiting for your reply.

First of all, the goal of this PR is to provide the ability to retry in the MQ retry queue within the Retry module.
Here are two changes made:
1. The `EventMeshProducer`, `ProducerManager`, and `SendMessageContext` of each protocol have been standardized. The reason for this change is that Retry should be a standard processing method, regardless of the specific underlying protocol, and the usage of these three protocols is completely consistent, so they are merged.

2. In RetryContext, when the user turns on the switch, the message is sent to the retry queue asynchronously to retry.

---

首先, 此PR的目标是在Retry模块内部提供于MQ重试队列来进行重试的能力.
这里做了两个改动:
1. 对于各协议的EventMeshProducer, ProducerManager, SendMessageContext进行了标准化, 此改动的原因是Retry应该是一个标准的处理方式, 而不关心具体底层的协议, 并且这三者在各协议的用法完全一致, 因此进行了合并处理.
2. 在RetryContext中, 在用户打开开关时, 支持将消息异步发送至重试队列中, 以达到重试的效果.

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/RetryContext.java
Comment (xwm1992) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/RetryContext.java:139]: > Send the message to the MQ retry queue for retry, more accurately RocketMQ. Because other mainstream MQ does not have a retry queue like RocketMQ, for example, RabbitMQ's retry queue is a regular queue that is only controlled through the `x-max-retries` attribute in the message header, and Kafka does not have a retry queue.
> 
> This also vaguely reminds me of the reason why PR #4179 was not merged, as if considering that this method can only be implemented for RocketMQ and is difficult to extend to other MQ. Is that the reason? @xwm1992
> 
> 将消息发给MQ的重试队列进行重试，更准确的说是RocketMQ。因为其他主流的MQ并没有像RocketMQ一样的重试队列，比如RabbitMQ的重试队列是普通队列，只是通过消息头的`x-max-retries`属性来控制、Kafka没有重试队列。 这也让我模糊地想起了没有合并PR #4179的原因，好像是考虑这种方式只能针对RocketMQ来实现，难以拓展到其他MQ。是这个原因吗？@xwm1992

是的，这个pr的主要修改点在于支持扩展不同的重试策略，将依赖存储的重试策略与eventmesh解耦

Comment (yanrongzhen) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/RetryContext.java]: I optimized this issue in a new round of commits.

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/RetryContext.java
Comment (yanrongzhen) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/RetryContext.java:110]: Used to control the number of consumption, since the retry-topic method is used for retrying, failure to control may result in unlimited repeated consumption.

---

用于控制消费次数, 由于使用了retry-topic的方式进行重试, 不进行控制可能会导致无限制重复消费.

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/RetryContext.java
Comment (yanrongzhen) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/RetryContext.java:113]: The semantics of this API is to specify the retry behavior of the Retryable class. I think for the retry class, there is no need to perceive whether the upper layer is retrying in memory or based on storage.

---

这个api的语义是指定Retryable类的重试行为, 我认为对于重试类而言, 不需要感知上层到底是通过内存重试还是基于存储的重试.

Comment (yanrongzhen) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/RetryContext.java:113]: In addition, I set the run method to final, which is invisible to its subclasses.

---

另外, run方法我设置为了final, 这样对于其子类是不可见的.

File: eventmesh-runtime/build.gradle
Comment (yanrongzhen) [reply eventmesh-runtime/build.gradle:76]: This is because the runtime needs to be loaded into a specific middleware SPI implementation, similar to the `build.gradle` file in the runtime module, lines 49 and 40.

---

这是因为runtime需要加载到具体中间件的SPI实现, 类似的做法在runtime模块中的`build.gradle`文件, 第49,40行也可以见到.

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/grpc/push/AbstractPushRequest.java
Comment (yanrongzhen) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/grpc/push/AbstractPushRequest.java:163]: This is for the RetryContext can be `rePut`, support the `Timeout` object back into the time wheel, so I reserved the imported parameter.

---

这个是为了在RetryContext当中可以进行rePut操作, 支持把Timeout对象重新放到时间轮里, 所以我预留了该入参.

File: eventmesh-runtime/build.gradle
Comment (yanrongzhen) [reply eventmesh-runtime/build.gradle:76]: > Waiting for your reply.

Optimized in a new commit.

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/grpc/push/AbstractPushRequest.java
Comment (yanrongzhen) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/grpc/push/AbstractPushRequest.java:163]: OK, I removed this method in the new commit.

File: eventmesh-runtime/build.gradle
Comment (yanrongzhen) [reply eventmesh-runtime/build.gradle:76]: I don't think this dependency should be removed. If it is removed, the SPI will not take effect when the user specifies storage = rocketmq.

---

我觉得不应该删除此依赖, 如果删除, 当用户指定了storage=rocketmq的时候, 将不会生效此SPI.

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/RetryContext.java
Comment (yanrongzhen) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/RetryContext.java:81]: The current design is to extend Retry's other SPIs in the future. You can understand that RetryStrategy is an SPI facade interface.

---

目前的设计是为了今后可以扩展Retry的其他SPI, 你可以理解为RetryStrategy是一个SPI门面接口.

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/grpc/push/AbstractPushRequest.java
Comment (yanrongzhen) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/grpc/push/AbstractPushRequest.java:163]: I personally think it's okay to keep it.

Comment (yanrongzhen) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/grpc/push/AbstractPushRequest.java:163]: > Could `Timeout` in the `doRun (timeout)`<-- `run (timeout)` reserved for `reput()` be removed ?
> 
> 那为`rePut()`预留的`doRun(Timeout)` <-- `run(Timeout)`中的`Timeout`是不是可以删掉了？

Also, the rePut method I think will be re-added soon in the future, and it is a common practice to resubmit to the hashed time wheel when the retry process fails.

---

并且, rePut方法我认为在未来很快会被重新添加回来, 在重试过程中产生失败时, 重新提交到时间轮是一个很常见的做法.

Comment (yanrongzhen) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/grpc/push/AbstractPushRequest.java:162]: Parameter has been removed.

File: eventmesh-runtime/build.gradle
Comment (yanrongzhen) [reply eventmesh-runtime/build.gradle]: I think you are right. I removed the dependency just now.

Comment (yanrongzhen) [reply eventmesh-runtime/build.gradle]: I rollbacked the dependency changes, and then i will resubmit it after we are done discussing it.

Comment (yanrongzhen) [reply eventmesh-runtime/build.gradle:76]: The core question may be whether the user needs to declare the retry-rocketmq dependency.

Comment (yanrongzhen) [reply eventmesh-runtime/build.gradle:76]: > I sincerely believe that there is still a need for another reviewer to express their views on this. @xwm1992. Because perhaps I misunderstood something. Thank you for your timely response.
> 
> 我真诚地觉得还是需要有另一位reviewer来就此说一下看法，@xwm1992 ，因为也许是我哪里理解得不对。感谢您及时的回应。

In fact, if the runtime does not show the addition of dependencies, it means that users need to load the dependencies of this SPI implementation by themselves. I am worried that it will create a greater burden for users.

---

其实如果runtime不显示加入依赖, 就代表了用户需要自行去加载这个SPI实现的依赖, 我担心对于用户会产生更大的负担,

Comment (yanrongzhen) [reply eventmesh-runtime/build.gradle:76]: > I sincerely believe that there is still a need for another reviewer to express their views on this. @xwm1992. Because perhaps I misunderstood something. Thank you for your timely response.
> 
> 我真诚地觉得还是需要有另一位reviewer来就此说一下看法，@xwm1992 ，因为也许是我哪里理解得不对。感谢您及时的回应。

And because the runtime is generally an independent server, users can only make this switch take effect through a plugin, which is obviously more troublesome.

---

并且由于runtime一般是一个独立的服务器, 这样用户又只能通过如plugin的方式才能让此开关生效, 显然比较麻烦.

**Comment by yanrongzhen** [comment]
@pandaapo 

I used the plugin method to solve the dependency problem discussed before, please review it again, thanks.

---

我使用了plugin方式来解决了之前讨论的依赖的问题, 帮忙再review一下, 谢谢

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/CommonConfiguration.java
Comment (yanrongzhen) [reply eventmesh-common/src/main/java/org/apache/eventmesh/common/config/CommonConfiguration.java:116]: Isolation is configured to allowing users to retry using standalone mode when the storage medium is not using standalone.

---

是的, 配置隔离是为了应对当存储介质不使用standalone时, 也允许用户使用standalone模式重试。

Comment (xwm1992) [reply eventmesh-common/src/main/java/org/apache/eventmesh/common/config/CommonConfiguration.java:116]: I understand that the original intention here is to use the `default` strategy. There is no need to reflect it as standalone. It can be marked as default. Others, such as rocketmq and rabbitmq, can have different retry strategies. `eventMeshRetryPluginType = “default”;` That’s fine. @pandaapo @yanrongzhen 

---
我理解这里的本意是走默认的策略，没必要体现是standalone，可以标识为default，其他例如rocketmq，rabbitmq，可以有不同的重试策略实现，`eventMeshRetryPluginType = “default”;` 这样就好。 @pandaapo @yanrongzhen

Comment (xwm1992) [reply eventmesh-common/src/main/java/org/apache/eventmesh/common/config/CommonConfiguration.java:116]: The relationship between the retry strategy plugin and storage can be reflected in the document, and there is no need to strongly bind the configuration name and code logic. I think everyone can understand this.

---
重试策略插件与使用的存储的关系，可以体现在文档中，不用在配置名称和代码逻辑上强绑定。这点我想大家也是都能理解的。

Comment (xwm1992) [reply eventmesh-common/src/main/java/org/apache/eventmesh/common/config/CommonConfiguration.java]: I understand that the current modification has nothing to do with `standalone`. If the `retry strategy` is configured, the corresponding configuration will be used. If the configuration is not loaded, the `default strategy` will be used. For the situation the rocketmq strategy is used, but the rabbitmq is used for storage. Just mark it in the documentation guide, or add a judgment to determine whether the strategy matches the storage, or throw a clear mismatch exception.

---
我理解目前的修改与`standalone` 没什么关系了，配置了`retry策略`就是走对应配置的，如果配置的没加载到就是走`default策略`，对于使用retry策略使用的`rocketmq策略`，但是存储使用的rabbitmq的这种情况，在文档指引中体现标识一下就行了，或者就是加一个策略与存储是否匹配的判断，抛出一个明确的不匹配的异常也可以。

Comment (yanrongzhen) [reply eventmesh-common/src/main/java/org/apache/eventmesh/common/config/CommonConfiguration.java]: The judgment of whether it matches needs to determine whether the strings of the two plug-in types are equal, which may lead to the coupling of the configuration.

---

是否匹配的判断需要判断两者插件类型的字符串是否相等, 感觉可能会导致配置的耦合。

Comment (xwm1992) [reply eventmesh-common/src/main/java/org/apache/eventmesh/common/config/CommonConfiguration.java]: Yes, there will be coupling, so either there are relevant descriptions in the document guidelines to make it clear to users, or coupling judgment processing must be carried out in the code. The judgment conditions are: if the retry strategy is used or the corresponding The storage module is the same, or it is the default strategy.

---
是的，是会存在耦合的情况，所以要么文档指引中有相关描述，让使用人清楚，要么就是要在代码中进行耦合的判断处理，判断的条件就是：如果使用了重试策略要么与对应的storage模块相同，要么就是default策略。

**Comment by yanrongzhen** [comment]
I resubmitted the code based on the latest code review, please help to review it again, thanks.

**Review by xwm1992 — APPROVED** [review]

**Pull request merged**
